### PR TITLE
Upgrade log4j to 2.16.0

### PR DESCRIPTION
**Pull request by arteam**
We upgraded log4j to 2.15.0 in #81709 and we can do the next upgrade. It makes JNDI lookups opt-in and also fixes [CVE-2021-45046](https://github.com/advisories/GHSA-7rjr-3q55-vv33)

This also reverts #81629 where we removed the `JndiLookup` class from the log4j jar.

### CODE REVIEW

**Comment by arteam** [comment]
@elasticmachine run elasticsearch-ci/part-1

**Review by mark-vieira — APPROVED** [review]
LGTM

**Review by pgomulka — CHANGES_REQUESTED** [review]
it looks good. I would prefer to split a revert and upgrade into separate PRs.
There should also be a revert of ef64808caf3902ea50014aebcd3026aa2ef4da0b before we upgrade to 2.16?

also it looks like there should be another log4j release very soon 
https://github.com/apache/logging-log4j2/commit/4456909#r61971917

File: distribution/build.gradle [hunk 1]
```diff
@@ -275,10 +275,6 @@ configure(subprojects.findAll { ['archives', 'packages'].contains(it.name) }) {
         }
       }
     }
-    all {
```
[the remark this answers is not in the document]
Comment (arteam) [reply distribution/build.gradle:278]: Let me extract the revert into a separate PR and then we can rebase this PR when the revert gets merged

**Comment by jasvinder1107** [comment]
@arteam  can you please look at why merging is blocked. This is high priority fix that we need to do  patch this before year end.

File: distribution/build.gradle (hunk 1 above)
[the remark this answers is not in the document]
Comment (arteam) [reply distribution/build.gradle:278]: Submitted as #81783

**Comment by jasvinder1107** [comment]
@arteam  thank you for so quick response. @pgomulka please give your blessing on this ps so this can be merged. :)

**Comment by jasvinder1107** [comment]
@arteam @pgomulka  I am sorry to being chatty here, can you please help us merging this asap ?

**Review by jasvinder1107 — APPROVED** [review]

**Comment by DJRickyB** [comment]
closes #81867

**Comment by ChrisHegarty** [comment]
We're already at 2.15.0 and have stripped JnidLookup.class, which is arguably more secure than 2.16.0. The stripping greatly outweighs the benefits of 2.16.0. The 2.16.0 release has a static dependency on JndiLookup.class. We're awaiting 2.6.1 where will be keep the stripping.

**Review by ChrisHegarty — CHANGES_REQUESTED** [review]
Please do not merge this PR, we're waiting for 2.16.1

**Comment by jasvinder1107** [comment]
@ChrisHegarty  so just to understand you correctly there will be stripping of the jndiclass from 2.16.1 also ?. Also can you please give us light on CVE in the 2.16.0 ?

**Comment by dionrhys** [comment]
Log4j2 2.17.0 has now been released with a fix to CVE-2021-45105, opt-in system properties for JNDI, and removing support for LDAP(S) over JNDI.

- https://github.com/apache/logging-log4j2/blob/6b1581901ba7a107cdc4a2208ecec03655722b44/RELEASE-NOTES.md
- https://logging.apache.org/log4j/2.x/security.html

**Comment by costin** [comment]
Superseded by https://github.com/elastic/elasticsearch/pull/81902 (Upgrade to Log4J 2.17.0)

**Pull request closed without merging**